docs/source/format/CDataInterface.rst

```diff
@@ -0,0 +1,773 @@
+.. distributed with this work for additional information
+.. regarding copyright ownership.  The ASF licenses this file
+.. to you under the Apache License, Version 2.0 (the
+.. "License"); you may not use this file except in compliance
+.. with the License.  You may obtain a copy of the License at
+
+..   http://www.apache.org/licenses/LICENSE-2.0
+
+.. Unless required by applicable law or agreed to in writing,
+.. software distributed under the License is distributed on an
+.. "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY
+.. KIND, either express or implied.  See the License for the
+.. specific language governing permissions and limitations
+.. under the License.
+
+==========================
+The Arrow C data interface
+==========================
+
+Rationale
+=========
+
+Apache Arrow is designed to be a universal in-memory format for the representation
+of tabular ("columnar") data. However, some projects may face a difficult
+choice between either depending on a fast-evolving project such as the
+Arrow C++ library, or having to reimplement adapters for data interchange,
+which may require significant, redundant development effort.
+
+The Arrow C data interface defines a very small, stable set of C definitions
+that can be easily *copied* in any project's source code and used for columnar
+data interchange in the Arrow format.  For non-C/C++ languages and runtimes,
+it should be almost as easy to translate the C definitions into the
+corresponding C FFI declarations.
+
+Applications and libraries can therefore work with Arrow memory without
+necessarily using Arrow libraries or reinventing the wheel. Developers can
+choose between tight integration
+with the Arrow *software project* (benefitting from the growing array of
+facilities exposed by e.g. the C++ or Java implementations of Apache Arrow,
+but with the cost of a dependency) or minimal integration with the Arrow
+*format* only.
+
+Goals
+-----
+
+* Expose an ABI-stable interface.
+* Make it easy for third-party projects to implement support for (including partial
+  support where sufficient), with little initial investment.
+* Allow zero-copy sharing of Arrow data between independent runtimes
+  and components running in the same process.
+* Match the Arrow array concepts closely to avoid the development of
+  yet another marshalling layer.
+* Avoid the need for one-to-one adaptation layers such as the limited
+  JPype-based bridge between Java and Python.
+* Enable integration without an explicit dependency (either at compile-time
+  or runtime) on the Arrow software project.
+
+Ideally, the Arrow C data interface can become a low-level *lingua franca*
+for sharing columnar data at runtime and establish Arrow as the universal
+building block in the columnar processing ecosystem.
+
+Non-goals
+---------
+
+* Expose a C API mimicking operations available in higher-level runtimes
+  (such as C++, Java...).
+* Data sharing between distinct processes.
```

Comment:
For clarity perhaps say "between distinct processes or serialization to disk in any way."

Also add

* Forward or backward compatibility guarantees; applications requiring such guarantees should use the IPC protocol

Reply:
Well, the interface is intended to be ABI-stable.

Reply:
Right but the compatibility guarantees made w.r.t. the Flatbuffers metadata are stronger, effectively we can change the "ABI" of the Flatbuffers messages (which would be like adding members to structs in C) but old readers will simply ignore new fields that it does not know about. If the struct sizes change in the C API it will break old readers